trace/report_buffer.go

```diff
@@ -109,40 +117,48 @@ func (buf *reportBuffer) addWitness(w *witnessWithInfo) {
 		return
 	}
 
-	buf.UploadReportsRequest.AddWitnessReport(witnessReport)
+	buf.activeUploadReport.AddWitnessReport(witnessReport)
 }
 
 func (buf *reportBuffer) Flush() error {
-	if buf.UploadReportsRequest.IsEmpty() {
+	if buf.activeUploadReport.IsEmpty() {
 		return nil
 	}
 
-	// Ensure the buffer is empty when we return.
-	defer buf.UploadReportsRequest.Clear()
-
-	// Upload to the back end.
-	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
-	defer cancel()
-
-	err := buf.collector.learnClient.AsyncReportsUpload(ctx, buf.collector.getLearnSession(), &buf.UploadReportsRequest)
-	if err != nil {
-		switch e := err.(type) {
-		case rest.HTTPError:
-			if e.StatusCode == http.StatusTooManyRequests {
-				// XXX Not all commands that call into this code have a --rate-limit
-				// option.
-				err = errors.Wrap(err, "your witness uploads are being throttled. Postman Insights will generate partial results. Try reducing the --rate-limit value to avoid this.")
+	// Cache current active upload report and reset it to next report.
+	// If witness rate is very high, reading from the channel could still block.
+	report := buf.activeUploadReport
+	learnSessions := buf.collector.getLearnSession()
+	buf.activeUploadReport = <-buf.uploadReports
+	go func() {
```

Comment:
I feel like it makes sense to swap the order of these.

If the other buffer is not yet free, why should we delay uploading the activeUploadReport?  Let's start the goroutine first, and then block.